fluent-merge/src/index.js

```diff
@@ -0,0 +1,66 @@
+import { Resource, parse, serialize } from 'fluent-syntax';
+
+/*
+ * Merge text files into one file.
+ */
+export function mergeFiles(...files) {
+  const resources = files.map(parse);
+  const merged = resources.reduce(merge);
+  return serialize(merged);
+}
+
+/*
+ * Merge AST trees into one tree.
+ */
+export function mergeResources(...resources) {
+  const merged = resources.reduce(merge);
+  return merged;
+}
+
+/*
+ * Merge messages of two resources.
+ *
+ * The first resource is treated as the reference. Sections and Junk entries
+ * are ignored.
+ */
+function merge(res1, res2) {
+  const body = [];
+
+  for (const message of messages(res1)) {
+    const other = findMessage(res2, message.id.name);
+    if (other) {
+      body.push(other);
+    } else {
+      body.push(message);
+    }
+  }
+
+  const merged = new Resource(body, res1.comment);
+  merged.source = serialize(merged);
+
+  return merged;
+}
+
+/*
+ * Find a message in `res` by its `id`.
+ */
+function findMessage(res, id) {
+  for (const other of messages(res)) {
+    if (other.id.name === id) {
+      return other;
+    }
+  }
+
+  return null;
+}
+
+/*
+ * Create an iterator over Resource's messages.
+ */
+function* messages(res) {
+  for (const entry of res.body) {
+    if (entry.type === 'Message') {
+      yield entry;
+    }
+  }
+}
```

Comment:
Due to the way you're using it (iterate over `res1`, for each entry in it iterate over `res2`) and the fact that in most cases I'd expect that most entries will be in both, I'd say that `let msgs = res.body.filter(...)` done once should be way cheaper memory wise and faster.
But this will work, so up to you:)